### PR TITLE
[WAM] Let MSALRuntime handle the no scope validation

### LINKED ISSUE #3675: [Request] WAM support for no scopes
**Logs and network traces**
none

**Which version of MSAL.NET are you using?**
latest

**Platform**
any

**Repro**

```csharp
var authResult = await pca.AcquireTokenInteractive(new[] { "" })
                                          .ExecuteAsync();
```

**Expected behavior**
Friendly exception

**Actual behavior**
Flow proceeds and results in unfriednly exceptions

Fix: authenticationRequestParameters.HasScopes returns true if empty string is passed as scopes

### PR DESCRIPTION
**Pull request by gladjohn**
Fixes #3675 

**Changes proposed in this request**
- Remove no scope check from MSAL .NET code 
- MSAL Runtime handles the no scope validation, catch the exception and throw a MSALServiceException 

**Testing**
Dev App

**Performance impact**
None

**Documentation**
- [ ] All relevant documentation is updated.

### CODE REVIEW

**Comment by pmaytak** [comment]
What is the reasoning for moving this check to MsalRuntime layer?

I think having it in MSAL is better - the error is more specific. Otherwise, MsalRuntime will throw a generic ApiContractViolation exception and the user will have to figure out what's wrong. Also I think it should stay a client exception, not service exception.

Can you remind me, what was the end result of the conversation about why MsalRuntime won't support empty (or oidc only) scopes?

**Comment by gladjohn** [comment]
> What is the reasoning for moving this check to MsalRuntime layer?
> 
> I think having it in MSAL is better - the error is more specific. Otherwise, MsalRuntime will throw a generic ApiContractViolation exception and the user will have to figure out what's wrong. Also I think it should stay a client exception, not service exception.
> 
> Can you remind me, what was the end result of the conversation about why MsalRuntime won't support empty (or oidc only) scopes?

Scopes are a required parameter. We need to pass scopes for MSAL CPP. This check was recently added on the C++ layer. Also there was a bug when we used magic scopes, it will let you sing in but error out during sign out. Now that has been fixed. If we decide to add the no scope check in MSAL .Net, we just need to check for empty / null scopes. The magic scope check we used to do needs to be removed as that scenario works now.

File: tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/RuntimeBrokerTests.cs [hunk 1]
```diff
@@ -239,6 +241,36 @@ public async Task WamListWindowsWorkAndSchoolAccountsAsync()
             var account = accounts.FirstOrDefault();
             Assert.AreEqual(labResponse.User.Upn, account.Username);
         }
+
+        [DataTestMethod]
+        [DataRow("")]
+        [DataRow(" ")]
+        [DataRow(null)]
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/RuntimeBrokerTests.cs:248]: Too expensive for integration test, pick one :)

File: tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/RuntimeBrokerTests.cs [hunk 2]
```diff
@@ -239,6 +241,36 @@ public async Task WamListWindowsWorkAndSchoolAccountsAsync()
...
+            Func<IntPtr> windowHandleProvider = () => intPtr;
+
+            IPublicClientApplication pca = PublicClientApplicationBuilder
+               .Create("04f0c124-f2bc-4f59-8241-bf6df9866bbd")
+               .WithAuthority("https://login.microsoftonline.com/organizations")
+               .WithParentActivityOrWindow(windowHandleProvider)
+               .WithBrokerPreview()
+               .Build();
+
+            var ex = await AssertException.TaskThrowsAsync<MsalServiceException>(
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/RuntimeBrokerTests.cs:266]: I was under the impression that MSAL C++ now allows no scopes?

Comment (gladjohn) [reply tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/RuntimeBrokerTests.cs:266]: Per discussion I had with @MSamWils and @logdog16, scopes is a required parameter for MSAL C++. But the magic scopes can be passed and token retrieved. There was a bug with magic scopes and signout but that is now fixed. @MSamWils can you please confirm the behavior

Comment (MSamWils) [reply tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/RuntimeBrokerTests.cs:266]: That's correct, passing the magic scope should works fine including sign out and sign back in.

Comment (pmaytak) [reply tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/RuntimeBrokerTests.cs:264]: I thought if the user passes empty scopes, we add [default/reserved](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/80c7824871bd6369e740479ba0497850bfbe51de/src/client/Microsoft.Identity.Client/OAuth2/OAuthConstants.cs#L101) scopes, like [here](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/80c7824871bd6369e740479ba0497850bfbe51de/src/client/Microsoft.Identity.Client/OAuth2/TokenClient.cs#L62). Seems like for old WAM, we check for null only, not empty, see [here](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/80c7824871bd6369e740479ba0497850bfbe51de/src/client/Microsoft.Identity.Client/Platforms/Features/WamBroker/AadPlugin.cs#L266).

So for MSAL runtime, we should pass default scopes, if user scopes are null/empty. Yes?

Comment (pmaytak) [reply tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/RuntimeBrokerTests.cs:264]: I thought if the user passes empty scopes, we add [default/reserved](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/80c7824871bd6369e740479ba0497850bfbe51de/src/client/Microsoft.Identity.Client/OAuth2/OAuthConstants.cs#L101) scopes, like [here](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/80c7824871bd6369e740479ba0497850bfbe51de/src/client/Microsoft.Identity.Client/OAuth2/TokenClient.cs#L62). Seems like for old WAM, we check for null only, not empty, see [here](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/80c7824871bd6369e740479ba0497850bfbe51de/src/client/Microsoft.Identity.Client/Platforms/Features/WamBroker/AadPlugin.cs#L266).

So for MSAL runtime, we should pass default scopes, if user scopes are null/empty. Yes?

**Review by pmaytak — CHANGES_REQUESTED** [review]

File: tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/RuntimeBrokerTests.cs (hunk 2 above)
Comment (gladjohn) [reply tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/RuntimeBrokerTests.cs:264]: sounds good, let me fix it

Comment (gladjohn) [reply tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/RuntimeBrokerTests.cs]: @pmaytak please review

File: src/client/Microsoft.Identity.Client.Broker/WamAdapters.cs [hunk 3]
```diff
@@ -165,14 +165,23 @@ private enum ResponseStatus
...
             var authParams = new NativeInterop.AuthParameters
                 (authenticationRequestParameters.AppConfig.ClientId,
                 authenticationRequestParameters.Authority.AuthorityInfo.CanonicalAuthority.ToString());
 
-            //scopes
-            authParams.RequestedScopes = string.Join(" ", authenticationRequestParameters.Scope);
+            //When no scopes are passed, add the default scopes
+            if (!ScopeHelper.HasNonMsalScopes(authenticationRequestParameters.Scope))
+            {
+                string defaultScopes = "email offline_access openid profile";
```
Comment (pmaytak) [inline src/client/Microsoft.Identity.Client.Broker/WamAdapters.cs:176]: We don't send email scope. Maybe create a `const string = "offline_access openid profile"` in OAuthConstants.
https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/a927f8c716677893c116333d6d8d2d12618f102b/src/client/Microsoft.Identity.Client/OAuth2/OAuthConstants.cs#L97-L99

File: tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/RuntimeBrokerTests.cs [hunk 4]
```diff
@@ -239,6 +241,43 @@ public async Task WamListWindowsWorkAndSchoolAccountsAsync()
...
+
+            }
+            catch (MsalUiRequiredException ex)
+            {
+                Assert.IsTrue(!string.IsNullOrEmpty(ex.ErrorCode));
+            }
+            catch (MsalServiceException)
+            {
+                Assert.Fail("MsalServiceException should not have been thrown.");
+            }
```
Comment (pmaytak) [inline tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/RuntimeBrokerTests.cs:279]:
```suggestion
            var ex = await AssertException.TaskThrowsAsync<MsalServiceException>(
                 () => pca.AcquireTokenSilent(new string[] { scopes }, PublicClientApplication.OperatingSystemAccount)
                        .ExecuteAsync())
                        .ConfigureAwait(false);

            Assert.IsFalse(string.IsNullOrEmpty(ex.ErrorCode));
```

**Review by pmaytak — APPROVED** [review]

**Review by pmaytak — APPROVED** [review]

**Pull request merged**